pkg/solana/client/multinode/transaction_sender.go

```diff
@@ -94,16 +91,26 @@ type TransactionSender[TX any, CHAIN_ID ID, RPC SendTxRPCClient[TX]] struct {
 // * If there is at least one terminal error - returns terminal error
 // * If there is both success and terminal error - returns success and reports invariant violation
 // * Otherwise, returns any (effectively random) of the errors.
-func (txSender *TransactionSender[TX, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) (SendTxReturnCode, error) {
-	txResults := make(chan sendTxResult)
-	txResultsToReport := make(chan sendTxResult)
+func (txSender *TransactionSender[TX, RESULT, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) RESULT {
+	txResults := make(chan RESULT)
+	txResultsToReport := make(chan RESULT)
 	primaryNodeWg := sync.WaitGroup{}
 
-	ctx, cancel := txSender.chStop.Ctx(ctx)
-	defer cancel()
+	if txSender.State() != "Started" {
+		return txSender.newResult(errors.New("TransactionSender not started"))
+	}
+
+	txSenderCtx, cancel := txSender.chStop.NewCtx()
+	reportWg := sync.WaitGroup{}
+	defer func() {
+		go func() {
+			reportWg.Wait()
+			cancel()
+		}()
+	}()
```

Comment:
This is incredibly unusual. Does `reportWg` only exist in order to coordinate this cancellation?

Reply:
Yeah, we need to ensure `cancel` is called after the concurrent call to `reportSendTxAnomalies`, as well as the entire `SendTransaction` function, returns.

Reply:
I only see one caller of Add/Done on `reportWg`. Why do we need to defer spawning a goroutine to wait for one thing? Why not just call `cancel()` inline from the other goroutine that calls `Done`?

Reply:
We're also waiting for `collectTxResults` to finish as well, which is why it's deferred. We can't call `cancel()` after `collectTxResults` directly nor after `reportSendTxAnomalies` since we must wait for both to complete.

Reply:
Why do they have to _complete_ before we cancel the context which is wired to the goroutines that fed them inputs? We already wait on those goroutines to close the channels some where else. Can we just cancel there?
```diff
		primaryNodeWg.Wait()
		close(txResultsToReport)
		close(txResults)
+		cancel()
```

Reply:
Is it guaranteed in this case that `collectTxResults` would read the last result before exiting due to `ctx.Done()`?

I see that `case r := <-txResults:` would be waiting for a result which would technically come before we call `cancel`, however does this guarantee that `ctx.Done()` does not trigger first?

Reply:
But `cancel` is associated with `txSenderCtx` which is not passed to `collectTxResults`.

Reply:
The initial idea was to call cancel when `reportSendTxAnomalies` - the last background job spawned within the call, is done. 
But as `reportSendTxAnomalies` ignores context we can simplify the code and cancel after `primaryNodeWg.Wait()`.
@DylanTinianov my bad, sorry for overcomplicating.